shared/src/main/scala/mlscript/JSBackend.scala

```diff
@@ -33,16 +34,27 @@ class JSBackend {
     * This function only translate name patterns and object patterns. I was thinking if we can
     * support literal parameter matching by merging multiple function `def`s into one.
     */
-  private def translatePattern(t: Term): JSPattern = t match {
+  private def translatePattern(t: Term)(implicit scope: Scope): JSPattern = t match {
     // fun x -> ... ==> function (x) { ... }
     // should returns ("x", ["x"])
-    case Var(name) => JSNamePattern(name)
+    case Var(name) => {
+      val runtimeName = scope.declareParameter(name)
+      JSNamePattern(runtimeName)
+    }
     // fun { x, y } -> ... ==> function ({ x, y }) { ... }
     // should returns ("{ x, y }", ["x", "y"])
     case Rcd(fields) =>
       JSObjectPattern(fields map {
-        case (Var(nme), (Var(als), _)) if nme === als => nme -> N
-        case (Var(nme), (subTrm, _))                  => nme -> S(translatePattern(subTrm))
+        case (Var(nme), (Var(als), _)) if nme === als => {
+          val runtimeName = scope.declareParameter(nme);
+          if (JSField.isValidIdentifier(nme)) nme -> N
```

Comment:
Similar as my previous comment: this logic seems brittle. What if `runtimeName` ended up different than `nme` here?

What we should test here is whether `runtimeName === als`, not whether `nme === als`! And we should _always_ use `runtimeName`.